src/containers.py

```diff
@@ -0,0 +1,188 @@
+                                    command,
+                                    stderr=True,
+                                    volumes=bind_paths,
+                                    working_dir=working_dir,
+                                    environment=[environment]).decode('utf-8')
+
+        # TODO does this cleanup need to still run even if there was an error in the above run command?
+        # On Unix, files written by the above Docker run command will be owned by root and cannot be modified
+        # outside the container by a non-root user
+        # Reset the file owner and the group inside the container
+        try:
+            # Only available on Unix
+            uid = os.getuid()
+            gid = os.getgid()
+
+            all_modified_volume_contents = set()
+            for src_path in pre_volume_contents.keys():
+                # Assumes the Docker run call is the only process that modified the contents
+                # Only considers files that were added, not files that were modified
+                post_volume_contents = set(src_path.iterdir())
+                modified_volume_contents = post_volume_contents - pre_volume_contents[src_path]
+                modified_volume_contents = [str(convert_docker_path(src_path, src_dest_map[src_path], p)) for p in
+                                            modified_volume_contents]
+                all_modified_volume_contents.update(modified_volume_contents)
+
+            # This command changes the ownership of output files so we don't
+            # get a permissions error when snakemake or the user try to touch the files
+            # Use --recursive because new directories could have been created inside the container
+            # Do not run the command if no files were modified
+            if len(all_modified_volume_contents) > 0:
+                chown_command = ['chown', f'{uid}:{gid}', '--recursive']
+                chown_command.extend(all_modified_volume_contents)
+                chown_command = ' '.join(chown_command)
+                client.containers.run(container,
+                                    chown_command,
+                                    stderr=True,
+                                    volumes=bind_paths,
+                                    working_dir=working_dir,
+                                    environment=[environment]).decode('utf-8')
+
+        # Raised on non-Unix systems
+        except AttributeError:
+            pass
+
+        # TODO: Not sure whether this is needed or where to close the client
+        client.close()
+
+    except Exception as err:
+        print(err)
+    # Removed the finally block to address bugbear B012
+    # "`return` inside `finally` blocks cause exceptions to be silenced"
+    # finally:
+    return out
+
+
+def run_container_singularity(container: str, command: List[str], volumes: List[Tuple[PurePath, PurePath]], working_dir: str, environment: str = 'SPRAS=True'):
```

Comment:
I just merged #66, which modified this function. We'll have to pull in those changes here.